rfcs/text/0019_lifecycle_preboot.md

```diff
@@ -0,0 +1,246 @@
+export interface InternalPrebootServiceSetup {
+  readonly waitUntilCanBoot: () => Promise<{ shouldReloadConfig: boolean }>;
+}
+```
+
+### 3.2.10 Bootstrap
+
+We'll update Kibana bootstrap sequence to include `preboot` stage and to conditionally reload configuration before proceeding to `setup` and `start` stages:
+
+```ts
+// NOTE(azasypkin): some functionality and checks have been omitted for brevity.
+const { preboot } = await root.preboot();
+
+const { shouldReloadConfig } = await preboot.waitUntilCanBoot();
+if (shouldReloadConfig) {
+  await reloadConfiguration('pre-boot request');
+}
+
+await root.setup();
+await root.start();
+```
+
+It's not yet clear if we need to adjust the base path proxy to account for this new lifecycle stage (see [unresolved question 3](#83-development-mode-and-basepath-proxy)).
+
+# 4. Drawbacks
+
+The main drawback is that proposed changes affect quite a few Kibana Core services that may impose a risk of breaking something in the critical parts of Kibana.
+
+# 5. Alternatives
+
+The most viable alternative to support interactive setup mode for Kibana was a standalone application that would be completely separated from Kibana. We ruled out this option since we won't be able to leverage existing and battle-tested Core services, UI components, and development tools. This would make the long-term maintenance burden unreasonably high.
+
+# 6. Adoption strategy
+
+The new `preboot` stage doesn't need an adoption strategy since it's intended for internal platform use only.
+
+# 7. How we teach this
+
+The new `preboot` stage shouldn't need much knowledge sharing since it's intended for internal platform use only and doesn't affect the primary plugins. All new services, methods, and contracts will be sufficiently documented in the code.
+
+# 8. Unresolved questions
+
+## 8.1 Lifecycle stage name
+
+Is `preboot` the right name for this new lifecycle stage? Do we have a better alternative?
+
+## 8.2 Core client-side changes
+
+The server-side part of the `preboot` plugins will follow a new `PrebootPlugin` interface that doesn't have a `start` lifecycle stage, but the client-side part will stay the same as for primary plugins. This significantly simplifies implementation and doesn't introduce any known technical issues, but, unfortunately, brings some inconsistency to the codebase. Namely, the client-side `CoreStart` contract, that `preboot` plugins will have access to, will still expose methods to work with the services that won't be available at the `preboot` stage (e.g., Saved Objects service). Does this tradeoff sound reasonable?
+
+## 8.3 Development mode and basepath proxy
+
+Currently, the base path proxy blocks any requests to Kibana until it receives `SERVER_LISTENING` message. Kibana's main process sends this message only after `start`, but we should change that to support interactive preboot applications. It's not yet clear how big the impact of this change will be.
```

Comment:
I never really understood the real value of this feature tbh. One option could be, when `isBootOnHold` returns true, to send this `SERVER_LISTENING` message when the preboot server is up.

Anyhow, this is only impacting local dev mode, so I don't think the impact should be significant, regardless of the solution we choose.

Reply:
> Anyhow, this is only impacting local dev mode, so I don't think the impact should be significant, regardless of the solution we choose.

Yep, agree, mentioned this just to have a more accurate picture on the amount of changes we'll need to make.